smartsim/_core/mli/infrastructure/storage/backbonefeaturestore.py

```diff
@@ -0,0 +1,399 @@
+        self.descriptor = descriptor
+        self.key = key
+
+    def __str__(self) -> str:
+        """Convert the event to a string
+
+        :returns: a string representation of this instance"""
+        return f"{str(super())}|{self.descriptor}|{self.key}"
+
+
+class EventPublisher(t.Protocol):
+    """Core API of a class that publishes events"""
+
+    def send(self, event: EventBase) -> int:
+        """The send operation"""
+
+
+class EventBroadcaster:
+    """Performs fan-out publishing of system events"""
+
+    def __init__(
+        self,
+        backbone: BackboneFeatureStore,
+        channel_factory: t.Optional[t.Callable[[str], CommChannelBase]] = None,
+        buffer_size: int = 100,
+    ) -> None:
+        """Initialize the EventPublisher instance
+
+        :param backbone: the MLI backbone feature store
+        :param channel_factory: factory method to construct new channel instances
+        :param buffer_size: Maximum number of events to store before discarding.
+        Discards oldest events first.
+        """
+        self._backbone = backbone
+        """The backbone feature store used to retrieve consumer descriptors"""
+        self._channel_factory = channel_factory
+        """A factory method used to instantiate channels from descriptors"""
+        self._channel_cache: t.Dict[str, t.Optional[CommChannelBase]] = defaultdict(
+            lambda: None
+        )
+        """A mapping of instantiated channels that can be re-used. Automatically 
+        calls the channel factory if a descriptor is not already in the collection"""
+        self._event_buffer: t.Deque[bytes] = deque(maxlen=buffer_size)
+        """A buffer for storing events when a consumer list is not found. Buffer size
+        can be fixed by passing the `buffer_size` parameter."""
+        self._num_discards: int = 0
+        """Number of messages discarded from the buffer since the last successful
+        broadcast"""
+        self._buffer_size: int = buffer_size
+        """Maximum size of the event buffer"""
+        self._descriptors: t.Set[str]
+        """Stores the most recent list of broadcast consumers. Updated automatically
+        on each broadcast"""
+
+    @property
+    def num_buffered(self) -> int:
+        """Return the number of events currently buffered to send"""
+        return len(self._event_buffer)
+
+    @property
+    def num_discarded(self) -> int:
+        """Return the number of events discarded from the buffer since the last send"""
+        return self._num_discards
+
+    def _save_to_buffer(self, event: EventBase) -> None:
+        """Places a serialized event in the buffer until full. Continues adding events
+        to the buffer using a first-in, first-discarded strategy.
+
+        :param event: The event to serialize and buffer"""
+        if len(self._event_buffer) == self._buffer_size:
+            self._event_buffer.popleft()
+            self._num_discards += 1
```

Comment:
Definitely worried about throwing away important messages. Do you think that a `buffer_size` of 100 is for sure large enough or do we need to experiment with that number a bit?